Feat/add data sources

mikesjcz opened the pull request.

Commit:
new connection to db by prisma in articleapp
remove fake data in articleapp
add serveces using db
new lib for PrismaClient
update README

Commit:
new prisma client for userapp
add new function (queries, mutations) in userapp and articleapp
remove fake data
connect database for apps of gateway
update README

Commit:
add dataSources to userApp and articleapp
add new testing datasources
add new graphql integration tests

kinapets commented on monorepo/monorepo-ws/apps/articleapp/src/data-sources/article-data-source.ts:
```diff
@@ -0,0 +1,27 @@
+    
+    async getAllArticles(): Promise<GqlArticle[]> {
+        const articles = await this.prisma.article.findMany();
+        console.log(articles);
```
🧹

kinapets commented on monorepo/monorepo-ws/apps/articleapp/src/data-sources/interfaces.ts:
```diff
@@ -0,0 +1,7 @@
+import { DataSource } from "apollo-datasource";
+import { GqlArticle, GqlMutationCreateArticleArgs } from "../resolvers/interfaces";
+
+export interface IArticleDataSource extends DataSource{
```
1️⃣ nepouzivej I cko u interfaces

2️⃣ prijde mi tady asi zbytecne na to mit interface (i kdyz ne rozhodne spatne) protoze ty data sourcy jsou dost konkretni - nicmen klido neodstranuj ale myslim ze budes mit trochu problem s namingem kvuli budu 1️⃣ = ale zkus se poprat, kdyztak poradim

kinapets commented on monorepo/monorepo-ws/apps/articleapp/src/main.ts:
```diff
@@ -1,5 +1,14 @@
+    article: IArticleDataSource
+}
+
+const port: number = Number(process.env.port);
```
ten port by mel vychazet z nejakeho configu urcite

mikesjcz replied:
Port vychází z .env projektu

kinapets commented on monorepo/monorepo-ws/apps/articleapp/src/resolvers/user/index.ts:
```diff
@@ -1,9 +1,17 @@
-    articles(user) {
-      return articles.filter((article) => article.authorId == user.id);
+    articles: async (user, __, c: Context) => {
+      return (await c.dataSources.article.getAllArticles()).filter(
```
❓tohle uz jde urcite resit ted lepe, preci nebude z databaze tahat vsechny articles a pak je filtrovat

kinapets commented on monorepo/monorepo-ws/apps/userapp/src/data-sources/interfaces.ts:
```diff
@@ -0,0 +1,11 @@
+import * as GqlInterfaces  from "../resolvers/interfaces";
+import {DataSource} from 'apollo-datasource'
+
+export interface IUserDataSource extends DataSource {
```
odstranit Icko a popracovat na namingu

kinapets commented on monorepo/monorepo-ws/apps/userapp/src/data-sources/user-data-source.ts:
```diff
@@ -0,0 +1,103 @@
+           }
+       })  
+       if(myUser == null){
+           throw new ApolloError("user doent exist!");        
```
typo

kinapets commented on monorepo/monorepo-ws/apps/userapp/src/data-sources/user-data-source.ts:
```diff
@@ -0,0 +1,103 @@
+               id: newFriendShip.friendWith
+           }        
+       })
+       if(user == null){
```
tohle se musi checkovat kdyz pouzivat findFirstOrThrow?

kinapets commented on monorepo/monorepo-ws/apps/userapp/src/data-sources/user-data-source.ts:
```diff
@@ -0,0 +1,103 @@
+           }        
+       })
+       if(user == null){
+           throw new ApolloError("friend doesn exist");        
```
typo

kinapets commented on monorepo/monorepo-ws/apps/userapp/src/data-sources/user-data-source.ts:
```diff
@@ -0,0 +1,103 @@
+    }
+    
+    async createdFriendShip(newFriendShip: GqlInterfaces.GqlMutationCreateFriendShipArgs): Promise<GqlInterfaces.GqlUser>{
+        const myUser = await this.prisma.user.findFirst({
```
proc myUser?

kinapets commented on monorepo/monorepo-ws/apps/userapp/src/data-sources/user-data-source.ts:
```diff
@@ -0,0 +1,103 @@
+           return this.AddFriendShip(myUser, newFriendShip)
+       }
+       const isFriend = myUser.friendOf.find(item => item.friendWithId == newFriendShip.friendWith)
+       if(typeof isFriend == 'undefined'){
```
isFriend evokuje ze to bude boolean ale to neni pravda 

zkus mozna pouzit metodu some

kinapets commented on monorepo/monorepo-ws/apps/userapp/src/data-sources/user-data-source.ts:
```diff
@@ -0,0 +1,103 @@
+                name: obj.name
+            }
+        })
+        if(user == null){
```
tohle checkovani neslo by to findFirstOrThrow?

kinapets commented on monorepo/monorepo-ws/libs/prisma-user-app-client/prisma/migrations/20220810125523_init/migration.sql:
```diff
@@ -0,0 +1,18 @@
+-- CreateTable
```
netreba mit `init` migraci 10x v ramci jednoho MR staci jedna migrace

Commit:
change IDataSource
change filter users articles
change findFirst to findFirstOrThrow
change find to some when check friendship
change User/ArticleDataSource to PrismaUser/ArticleDataSource
